### ISSUE TITLE
Allow Azure ARM instances to Join Domains

### ISSUE THREAD

**Issue by phinze** [hashicorp/terraform#5820]
Looks like this is functionality we had in ASM that we don't yet have in ARM.

Here is a blog post that describes the API - I think! :smile: 

https://blogs.msdn.microsoft.com/igorpag/2016/01/25/azure-arm-vm-domain-join-to-active-directory-domain-with-joindomain-extension/

**Comment by phinze** [comment]
Ping to @stack72 :heartbeat:

**Comment by stack72** [comment]
@phinze so i have been looking at how we would do this in AzureRM. Currently, the only think the AzureRM SDK supports is as follows:

```
{
  "id":"/subscriptions/########-####-####-####-############/resourceGroups/{resourceGroupName}/providers/Microsoft.Compute/virtualMachines/{virtualMachineName}/extensions/{extensionName}",
  "name": "extensionName",
  "type": "Microsoft.Compute/virtualMachines/extensions",
  "location": "westus",
  "properties": {
    "publisher": "Microsoft.OSTCExtensions",
    "type": "CustomScriptForLinux",
    "typeHandlerVersion": "1.2",
    "settings": {
      "fileUris": [ "scriptFilePpath"],
      "commandToExecute": "command"
    }
  }
}
```

The user would still need to provide the script that joins the machine to the domain.

**Comment by stevendborrelli** [comment]
Any idea how this would be implemented? We just ran into this as a roadblock.

**Comment by thebrianlopez** [comment]
+1

**Comment by thebrianlopez** [comment]
I was able to find a possible workaround to apply after the machine is provisioned using the azure cli. If we can execute this inline with terraform then we can bootstrap the resource. End game should be to have tf do this natively. Gonna see if i can hack something together for a PR when i have time.

```
azure vm extension set --resource-group SAMPLE010 \
    --vm-name sample010 \
    --name CustomScriptExtension \
    --publisher-name Microsoft.Compute \
    --version 1.7 \
    --reference-name Microsoft.Compute/virtualMachines/extensions \
    --public-config "{\"fileUris\": [\"https://iaasv2tempstoreeastus.blob.core.windows.net/vmextensionstemporary-00034001aaf10ad1-20160804045452853/script.ps1?sv=2015-04-05&sr=c&sig=fDVCUg%2FpwPPr0Bbwo5ssfMpZUqqwGsezjjzIzH6oIpM%3D&se=2016-08-05T04%3A54%3A52Z&sp=rw\"],\"commandToExecute\": \"powershell -exec Unrestricted -nopro -File script.ps1  \"}"
```

**Comment by pmcatominey** [comment]
This will be solved via #9962
